TombEngine/Objects/TR4/Entity/tr4_setha.cpp

```diff
@@ -494,182 +467,169 @@ namespace TEN::Entities::TR4
+				if (item->Animation.ActiveState != SETH_STATE_JUMP_PROJECTILE_ATTACK)
 				{
-					if (item->Animation.ActiveState <= SETH_STATE_BIG_PROJECTILE)
+					if (item->Animation.ActiveState <= SETH_STATE_SINGLE_PROJECTILE_ATTACK)
 					{
-						if (abs(height4 - item->Pose.Position.y) >= SECTOR(0.5f))
-						{
-							item->Animation.AnimNumber = Objects[item->ObjectNumber].animIndex + SETH_ANIM_IDLE_GET_SHOT;
-							item->Animation.ActiveState = SETH_STATE_GET_SHOT;
-							item->Animation.TargetState = SETH_STATE_GET_SHOT;
-						}
+						if (abs(height4 - item->Pose.Position.y) >= BLOCK(1 / 2.0f))
+							SetAnimation(item, SETH_ANIM_SOFT_RECOIL);
 						else
-						{
-							item->Animation.AnimNumber = Objects[item->ObjectNumber].animIndex + SETH_ANIM_GET_HEAVY_SHOT_SOMERSAULT;
-							item->Animation.ActiveState = SETH_STATE_GET_HEAVY_SHOT;
-							item->Animation.TargetState = SETH_STATE_GET_HEAVY_SHOT;
-						}
+							SetAnimation(item, SETH_ANIM_HARD_RECOIL_START);
 					}
 					else
 					{
-						item->Animation.AnimNumber = Objects[item->ObjectNumber].animIndex + SETH_ANIM_FLY_GET_HEAVY_SHOT;
-						item->Animation.ActiveState = SETH_STATE_FLY_GET_HEAVY_SHOT;
-						item->Animation.TargetState = SETH_STATE_FLY_GET_HEAVY_SHOT;
+						SetAnimation(item, SETH_ANIM_HOVER_RECOIL);
 					}
-
-					item->Animation.FrameNumber = g_Level.Anims[item->Animation.AnimNumber].frameBase;
 				}
 			}
 		}
 
 		CreatureAnimation(itemNumber, angle, 0);
 	}
 
-	void SethaAttack(int itemNumber)
+	void SethAttack(int itemNumber)
 	{
 		auto* item = &g_Level.Items[itemNumber];
 
 		item->ItemFlags[0]++;
 
-		auto pos1 = GetJointPosition(item, SethaAttack1.meshNum, Vector3i(SethaAttack1.Position));
-		auto pos2 = GetJointPosition(item, SethaAttack2.meshNum, Vector3i(SethaAttack2.Position));
-		int sR = 64;
-		int sG = (GetRandomControl() & 0x7F) + 64;
-		int sB = sG;
-		auto sparkColor = Vector3(sR, sG, sB);
-		int fR = 0;
-		int fG = (GetRandomControl() & 0x7F) + 64;
-		int fB = fG - 32;
-		auto flameColor = Vector3(fR, fG, fB);
+		auto pos1 = GetJointPosition(item, SethAttack1.meshNum, Vector3i(SethAttack1.Position));
+		auto pos2 = GetJointPosition(item, SethAttack2.meshNum, Vector3i(SethAttack2.Position));
+
+		int sparkR = 64;
+		int sparkG = Random::GenerateInt(64, 192);
+		int sparkB = sparkG;
+		auto sparkColor = Vector3(sparkR, sparkG, sparkB);
 
-		int size;
+		int flameR = 0;
+		int flameG = Random::GenerateInt(64, 192);
+		int flameB = flameG - 32;
+		auto flameColor = Vector3(flameR, flameG, flameB);
+
+		int scale = 0;
 
 		switch (item->Animation.ActiveState)
 		{
-		case SETH_STATE_SHOOT_RIGHT_AND_LEFT:
-		case SETH_STATE_FLY_AIR_SHOOT:
-			if (item->ItemFlags[0] < 78 && (GetRandomControl() & 0x1F) < item->ItemFlags[0])
+		case SETH_STATE_DUAL_PROJECTILE_ATTACK:
+		case SETH_STATE_HOVER_PROJECTILE_ATTACK:
+			if (item->ItemFlags[0] < 78 &&
+				(GetRandomControl() & 0x1F) < item->ItemFlags[0])
```

Comment:
Why this GetRandomControl was not replaced?

Reply:
I suppose it generates an int in the range [0, 32], but I don't know what `ItemFlags[0]` contains exactly.

Reply:
From the looks of it, it seems like some kind of AI modifier, kind of "frustration" with the attack. When `ItemFlags[0]` reaches certain value, a decision is made on certain kinds of attacks, after which his "frustration" is "resolved" (value is reset to 0).